### ISSUE TITLE
HTTP in prod mode

### ISSUE THREAD

**Issue by johnjelinek** [Caiyeon/goldfish#57]
I want to put goldfish behind a load balancer (haproxy). HAProxy already has the necessary TLS certs. Can you please support running goldfish in HTTP in prod mode?

**Comment by Caiyeon** [comment]
TL;DR:
1. Goldfish is fast enough to not require load balancing
2. Goldfish stores data in cubbyhole, so it cannot be load balanced
3. There are better ways to achieve 24/7 availability

---

John,

Goldfish is light-weight enough to not require load balancing. The live demo is setup with a heavily over-provisioned 512MB linux box, and it's running both goldfish and vault at the same time. CPU and storage usage rarely spikes enough for me to worry.

As well, having multiple goldfishes in the fish bowl (heh.) creates persistence problems. The policy change requests are stored in a goldfish server's vault token's cubbyhole. Because at the moment goldfish strictly launches from approle, no two instances of goldfish should ever operate with the same vault token. This means that for the user trying to pull up a policy change request, he has a chance of hitting the goldfish instance that does not have the change request in its cubbyhole.

The usage of cubbyhole will continue, because unseal tokens are ttl-wrapped, and if they are stored anywhere other than cubbyhole in vault, there's a chance that a privileged vault user can unwrap to receive unseal tokens. At the moment, the way goldfish guarantees that no one can unwrap these unseal tokens is because if goldfish's memory is locked down, no one can see the wrapping token, and they are guaranteed to expire within the hour if goldfish is not around to cleanup.

---

If you are trying to load balance to achieve high availability (and not performance gain), you have a couple of better options:
1. I would highly recommend setting up a deploy script to achieve one-click deployment. The deployment script should be *really* simple to automate.
2. If you truly want to achieve 24/7 availability, you can setup a systemd service, with a vault token privileged to generate approle tokens, and re-launch goldfish process whenever it dies (from excess load, or otherwise). Just be careful when setting it up with that token.
3. Alternatively, maintain 2-3 separate instances of goldfish at all times (but *without* load balancing). This way, when the primary goes down, the secondary can be still reached. It's just that the secondary one will not remember any policy change requests. But at least this way, the user will explicitly know why it is not able to remember the change request, as opposed to randomly not being able to bring it up (in the case of a load balancer)

**Comment by johnjelinek** [comment]
In my scenario, I don't have access to the HTTPS certs, those are managed
within HAProxy. I would like to do TLS termination from there and speak to
goldfish on an HTTP backend. Is that not going to be possible?

**Comment by Caiyeon** [comment]
I'll look into this. While it is theoretically a possibility, anytime settings restrictions are 'loosened', it is a bit of a concern.

**Comment by Caiyeon** [comment]
I will follow vault's convention of allowing HTTP communication (with TLS disabled) via a `tls_disable` flag: https://www.vaultproject.io/docs/configuration/listener/tcp.html

To that end, I do have to re-wire server.go a tiny bit, and update the documentation.
This will be available on master in a couple of days, and should be included in the 0.3.3 binaries (i.e. within a week)

Cheers

**Comment by johnjelinek** [comment]
Huzzah \0/

**Comment by Caiyeon** [comment]
Added in https://github.com/Caiyeon/goldfish/releases/tag/v0.3.3

**Comment by Caiyeon** [comment]
There was a problem with the `tlsDisable` arg logic. This has been fixed in https://github.com/Caiyeon/goldfish/commit/127504f5f65e1afaaca8bacc9e0196d35fa99a1f

And the release has been updated:
https://github.com/Caiyeon/goldfish/releases/tag/v0.3.3
